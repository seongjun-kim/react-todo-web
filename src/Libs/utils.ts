import React from 'react';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

export const renderWithRouter = (ui: React.ReactElement, { route = '/' } = {}) => {
  window.history.pushState({}, '', route);
  return render(ui, { wrapper: BrowserRouter });
};
