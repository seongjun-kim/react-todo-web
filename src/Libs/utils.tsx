import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter, BrowserRouter } from 'react-router-dom';
import { ToDoListProvider } from 'Contexts';
import App from 'App';

/**
 * [renderWithRouter]
 * useLocation과 같은 라우터 관련 훅을 포함한 단일 컴포넌트 테스트를 위한 렌더링 함수
 * 컴포넌트 테스트를 위한 함수이므로 Components 폴더 아래 테스트에 활용
 */
export const renderWithRouter = (ui: React.ReactElement, { route = '/' } = {}) => {
  window.history.pushState({}, '', route);
  return render(ui, { wrapper: BrowserRouter });
};

/**
 * [renderWithMemoryRouter]
 * 실제 앱에서의 구동 방식 테스트를 위한 렌더링 함수
 * 페이지 전환 테스트를 위한 함수이므로 Pages 폴더 아래 테스트에 활용
 *
 *  - initialEntries를 window.history 처럼 활용 가능
 *  - TestComponent를 활용한 앱 내 추가 컴포넌트 적용 테스트 가능
 *    - 주로 useLocation().pathname을 파악하는 용도로 활용
 */
export const renderWithMemoryRouter = (
  { initialIndex = 0, initialEntries = ['/'] } = {},
  TestComponent?: JSX.Element | JSX.Element[],
) => {
  return render(
    <MemoryRouter initialEntries={initialEntries} initialIndex={initialIndex}>
      {TestComponent}
      <ToDoListProvider>
        <App />
      </ToDoListProvider>
    </MemoryRouter>,
  );
};
