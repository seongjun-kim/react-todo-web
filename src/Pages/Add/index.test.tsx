import React from 'react';
import { Router, useLocation } from 'react-router-dom';
import { fireEvent, render, screen } from '@testing-library/react';
import 'jest-styled-components';

import { Add } from './index';
import { renderWithRouter } from 'Libs/utils';
import { ToDoListProvider } from 'Contexts';

describe('<Add />', () => {
  it('renders component correctly', () => {
    const { container } = renderWithRouter(<Add />, { route: '/' });

    const input = screen.getByPlaceholderText('Enter what to do...');
    expect(input).toBeInTheDocument();
    const button = screen.getByText('Add');
    expect(button).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('redirects to the root page after adding a new ToDo', () => {
    const TestComponent = () => {
      const { pathname } = useLocation();
      return (
        <ToDoListProvider>
          <div>{pathname}</div>
          <Add />
        </ToDoListProvider>
      );
    };
    renderWithRouter(<TestComponent />, { route: '/add' });

    const pathName = screen.getByText('/add');
    expect(pathName).toBeInTheDocument();

    const input = screen.getByPlaceholderText('Enter what to do...');
    const button = screen.getByText('Add');
    const sampleData = ['New ToDo'];

    fireEvent.change(input, { target: { value: sampleData[0] } });
    fireEvent.click(button);

    expect(pathName.textContent).toBe('/');
    expect(localStorage.getItem('ToDoList')).toBe(JSON.stringify(sampleData));
  });
});
