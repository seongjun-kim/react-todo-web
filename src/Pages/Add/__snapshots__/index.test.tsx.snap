// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Add /> renders component correctly 1`] = `
.c3 {
  text-align: center;
  background-color: #304ffe;
  padding: 10px 20px;
  border-radius: 8px;
  cursor: pointer;
}

.c3:hover {
  background-color: #1e40ff;
}

.c3:active {
  box-shadow: inset 5px 5px 10px rgba(0,0,0,0.2);
}

.c4 {
  color: #fff;
  font-size: 16px;
}

.c2 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin-right: 10px;
  font-size: 16px;
  padding: 10px 10px;
  border-radius: 8px;
  border: 1px solid #bdbdbd;
  outline: none;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: #fff;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 20px;
  border-radius: 8px;
  box-shadow: 5px 5px 10px rgba(0,0,0,0.2);
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <input
        class="c2"
        placeholder="Enter what to do..."
        value=""
      />
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          Add
        </div>
      </div>
    </div>
  </div>
</div>
`;
